From da3241a593d2a06ed3bd5cd76a01655c28da7695 Mon Sep 17 00:00:00 2001
From: Jens Rottmann <JRottmann@LiPPERTembedded.de>
Date: Wed, 8 Jan 2014 15:36:53 +0100
Subject: Ethernet is subtly different.

- Correct RGMII_nRST GPIO polarity.
- Atheros recommended adding a 22 Ohm serial resistor to RGMII_TXCLK,
  however measurement shows a smaller value is slightly better. Instead of
  another BOM change simply reduce the resistor at the i.MX6 output (i.e.
  increase drive strength) from 37 to 31 Ohm.
- We use pad GPIO16 for something else, no ENET_REF_CLK. A note in the
  SabreSD reference schematics says "This pin [GPIO_16] must be unconnected
  for Ethernet 1588 (time stamp) [Precision Time Protocol, PTP]
  functionality to work," but the smaller AR8035 doesn't support PTP
  anyway.
- Re-add PHY reset GPIO pinmux; was in the 'hog' group originally.
- RGMII_INT is connected exactly like on Freescale's SabreSD board, but
  their device tree file doesn't mention it anywhere so we won't either.
  With this signal Wake-on-LAN could be handled by the PHY alone, allowing
  to save more power by putting the i.MX6 MAC to sleep, however Freescale's
  driver doesn't support this and keeps the MAC awake for Wake-on-LAN.
- Document how ERR006687 workaround can be enabled, instead of having an
  unused, unexplained pinctrl group.
- We have longer RGMII traces and (since A4) a serial resistor in TXC. This
  seems to have shifted the signal slightly, the GTX_CLK delay register in
  the phy must be changed from 2.4 (default) to 1.3 ns so the clock edge
  hits the data bits properly again.

diff --git a/arch/arm/boot/dts/lec-imx6.dtsi b/arch/arm/boot/dts/lec-imx6.dtsi
index 048a7db..6907c29 100644
--- a/arch/arm/boot/dts/lec-imx6.dtsi
+++ b/arch/arm/boot/dts/lec-imx6.dtsi
@@ -343,7 +343,7 @@
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_enet>;
 	phy-mode = "rgmii";
-	phy-reset-gpios = <&gpio1 25 0>;
+	phy-reset-gpios = <&gpio1 25 GPIO_ACTIVE_LOW>; /* RGMII_nRST */
 	fsl,magic-packet;
 	status = "okay";
 };
@@ -720,7 +720,7 @@
 			fsl,pins = <
 				MX6QDL_PAD_ENET_MDIO__ENET_MDIO		0x1b0b0
 				MX6QDL_PAD_ENET_MDC__ENET_MDC		0x1b0b0
-				MX6QDL_PAD_RGMII_TXC__RGMII_TXC		0x1b0b0
+				MX6QDL_PAD_RGMII_TXC__RGMII_TXC		0x1b0b8 /* serial resistor added */
 				MX6QDL_PAD_RGMII_TD0__RGMII_TD0		0x1b0b0
 				MX6QDL_PAD_RGMII_TD1__RGMII_TD1		0x1b0b0
 				MX6QDL_PAD_RGMII_TD2__RGMII_TD2		0x1b0b0
@@ -733,13 +733,12 @@
 				MX6QDL_PAD_RGMII_RD2__RGMII_RD2		0x1b0b0
 				MX6QDL_PAD_RGMII_RD3__RGMII_RD3		0x1b0b0
 				MX6QDL_PAD_RGMII_RX_CTL__RGMII_RX_CTL	0x1b0b0
-				MX6QDL_PAD_GPIO_16__ENET_REF_CLK	0x4001b0a8
-			>;
-		};
-
-		pinctrl_enet_irq: enetirqgrp {
-			fsl,pins = <
-				MX6QDL_PAD_GPIO_6__ENET_IRQ		0x000b1
+				MX6QDL_PAD_ENET_CRS_DV__GPIO1_IO25   0x80000000 /* RGMII_nRST */
+				//MX6QDL_PAD_ENET_RXD1__GPIO1_IO26   0x80000000 /* RGMII_INT */
+				/* For the workaround improving network latency described in i.MX6
+				 * erratum 6687 disable the power button, re-mux as follows and
+				 * add "interrupts-extended" as in imx6q-sabresd-enetirq.dts. */
+				//MX6QDL_PAD_GPIO_6__ENET_IRQ		0x000b1
 			>;
 		};
 
diff --git a/arch/arm/mach-imx/mach-imx6q.c b/arch/arm/mach-imx/mach-imx6q.c
index 77441a0..be66ffb 100644
--- a/arch/arm/mach-imx/mach-imx6q.c
+++ b/arch/arm/mach-imx/mach-imx6q.c
@@ -156,7 +156,6 @@ static int ar8035_phy_fixup(struct phy_device *dev)
 	phy_write(dev, 0xd, 0x3);
 	phy_write(dev, 0xe, 0x805d);
 	phy_write(dev, 0xd, 0x4003);
-
 	val = phy_read(dev, 0xe);
 	phy_write(dev, 0xe, val & ~(1 << 8));
 
@@ -169,6 +168,15 @@ static int ar8035_phy_fixup(struct phy_device *dev)
 	 */
 	ar8031_phy_fixup(dev);
 
+	/* The LEC-iMX6 has longer traces and serial resistors in CLK. */
+	if (of_machine_is_compatible("adlink,lec-imx6")) {
+		/* Change GTX_CLK delay from 2.4 (default) to 1.3 ns. */
+		phy_write(dev, 0x1D, 0x0B);
+		val = phy_read(dev, 0x1E);
+		val = (val & ~0x0060) | 0x0020;
+		phy_write(dev, 0x1E, val);
+	}
+
 	/*check phy power*/
 	val = phy_read(dev, 0x0);
 	if (val & BMCR_PDOWN)
-- 
2.1.4

